### PR TITLE
Extend msvc (DLL) configuration to clang-cl and mingw

### PR DESCRIPTION
**Pull request by dg0yt**
This PR is based on patching in vcpkg. Originally contributed to enable builds with clang-cl, I modified it to cover mingw builds. This change reduced the C++ import lib on mingw by one third.

### CODE REVIEW

**Comment by dg0yt** [comment]
~~~
2022-11-22T06:57:22.4721616Z In file included from D:/a/geos/geos/include/geos/algorithm/PointLocation.h:24,
2022-11-22T06:57:22.5346164Z                  from D:/a/geos/geos/src/algorithm/ConvexHull.cpp:22:
2022-11-22T06:57:22.6752542Z D:/a/geos/geos/include/geos/geom/Location.h:32:32: error: 'dllexport' attribute ignored [-Werror=attributes]
2022-11-22T06:57:22.7377914Z    32 | enum class GEOS_DLL Location : char {
2022-11-22T06:57:22.8897398Z       |                                ^~~~
2022-11-22T06:57:23.0675800Z cc1plus.exe: all warnings being treated as errors
2022-11-22T06:57:23.1946999Z make[2]: *** [CMakeFiles/geos.dir/build.make:146: CMakeFiles/geos.dir/src/algorithm/ConvexHull.cpp.obj] Error 1
~~~
Hm, I saw this warning while testing, but I didn't build with `-Werror=attributes`. Not sure how to properly DLL-export `enum class` with mingw.

File: CMakeLists.txt [hunk 1]
```diff
@@ -187,7 +187,7 @@ target_compile_features(geos_cxx_flags INTERFACE cxx_std_11)
 target_compile_options(geos_cxx_flags INTERFACE
 	"$<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>>:-ffp-contract=off>"
 	"$<$<CXX_COMPILER_ID:GNU>:-ffp-contract=off>"
-	"$<$<CXX_COMPILER_ID:MSVC>:/fp:precise>"
+	"$<$<BOOL:${MSVC}>:/fp:precise>"
```
Comment (pramsey) [inline CMakeLists.txt:190]: Why the bool thing? Is the old way broken? (I find the old one more straightforward in my head.)

Comment (pramsey) [reply CMakeLists.txt:190]: Nevermind, you're picking off mingw and msvc with one shot, got it. Add a comment to that effect?

Comment (dg0yt) [reply CMakeLists.txt:190]: Unrelated to mingw. AFAIU `$<CXX_COMPILER_ID:MSVC>` doesn't cover clang-cl, but `${MSVC}` does.

**Pull request merged**